src/Blob.mo

```diff
@@ -1,53 +1,199 @@
+/// If you would like to manipulate Blobs, it is recommended that you convert
+/// Blobs to `[var Nat8]` or `Buffer<Nat8>`, do the manipulation, then convert back.
+///
+/// Import from the base library to use this module.
+/// ```motoko name=import
+/// import Blob "mo:base/Blob";
+/// ```
+///
+/// Some built in features not listed in this module:
+/// ```motoko include=import
+/// import Debug "mo:base/Debug";
+/// import Nat8 "mo:base/Nat8";
+///
+/// let blob = "\00\00\00\ff" : Blob; // blob literals, where each byte is delimited by a back-slash and represented in hex
+/// let numBytes = blob.size(); // => 4 (returns the number of bytes in the Blob)
+/// for (byte : Nat8 in blob.vals()) { // iterator over the Blob
+///   Debug.print(Nat8.toText(byte))
+/// }
+/// ```
 import Prim "mo:⛔";
 module {
-
-  /// An immutable, possibly empty sequence of bytes.
-  /// Given `b : Blob`:
-  ///
-  /// * `b.size() : Nat` returns the number of bytes in the blob;
-  /// * `b.vals() : Iter.Iter<Nat8>` returns an iterator to enumerate the bytes of the blob.
-  ///
-  /// (Direct indexing of Blobs is not yet supported.)
   public type Blob = Prim.Types.Blob;
+  /// Creates a `Blob` from an array of bytes (`[Nat8]`), by copying each element.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// let bytes : [Nat8] = [0, 255, 0];
+  /// let blob = Blob.fromArray(bytes); // => "\00\FF\00"
+  /// ```
+  public func fromArray(bytes : [Nat8]) : Blob = Prim.arrayToBlob bytes;
 
-  /// Returns a (non-cryptographic) hash of 'b'
-  public let hash : (b : Blob) -> Nat32 = Prim.hashBlob;
-
-  /// Returns `x == y`.
-  public func equal(x : Blob, y : Blob) : Bool { x == y };
-
-  /// Returns `x != y`.
-  public func notEqual(x : Blob, y : Blob) : Bool { x != y };
-
-  /// Returns `x < y`.
-  public func less(x : Blob, y : Blob) : Bool { x < y };
+  /// Creates a `Blob` from a mutable array of bytes (`[Nat8]`), by copying each element.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// let bytes : [var Nat8] = [var 0, 255, 0];
+  /// let blob = Blob.fromArrayMut(bytes); // => "\00\FF\00"
+  /// ```
+  public func fromArrayMut(bytes : [var Nat8]) : Blob = Prim.arrayMutToBlob bytes;
 
-  /// Returns `x <= y`.
-  public func lessOrEqual(x : Blob, y : Blob) : Bool { x <= y };
+  /// Converts a `Blob` to an array of bytes (`[Nat8]`), by copying each element.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// let blob = "\00\FF\00" : Blob;
+  /// let bytes = Blob.toArray(blob); // => [0, 255, 0]
+  /// ```
+  public func toArray(blob : Blob) : [Nat8] = Prim.blobToArray blob;
 
-  /// Returns `x > y`.
-  public func greater(x : Blob, y : Blob) : Bool { x > y };
+  /// Converts a `Blob` to a mutable array of bytes (`[Nat8]`), by copying each element.
+  ///
+  /// Example:
+  /// ```motoko include=import
+  /// let blob = "\00\FF\00" : Blob;
+  /// let bytes = Blob.toArrayMut(blob); // => [0, 255, 0]
```

Comment:
```suggestion
  /// let bytes = Blob.toArrayMut(blob); // => [var 0, 255, 0]
```

Reply:
(Is that how we display mutable arrays?)

Reply:
This is how it comes out in the interpreter:

`[0, 255, 0] : [var Nat8]`

Reply:
But maybe that's confusing. I'll use `[var 0, 255, 0]` instead

Reply:
Oh, that's probably because the value printer for the interpreter isn't type directed (as it should be)